File: backend/domain_monitor/config_loader.py
from __future__ import annotations

import json
import logging
import os
import threading
from pathlib import Path
from typing import Any

logger = logging.getLogger(__name__)

CONFIG_DIR = Path(
    os.environ.get(
        "DOMAIN_MONITOR_CONFIG",
        str(Path(__file__).resolve().parent / "config"),
    )
)

_cache: dict[str, Any] = {}
_lock = threading.Lock()


def load(name: str) -> dict[str, Any]:
    """Read a JSON config file once and memoise it.

    Config is data, not code: editing these files changes classification and
    scoring without touching a single module.
    """
    with _lock:
        if name in _cache:
            return _cache[name]
        path = CONFIG_DIR / name
        try:
            data = json.loads(path.read_text(encoding="utf-8"))
        except (OSError, ValueError) as exc:
            logger.error("Could not read config %s: %s", path, exc)
            data = {}
        _cache[name] = data
        return data


def topics() -> dict[str, list[str]]:
    return load("seo_topics.json").get("topics", {})


def spam_categories() -> dict[str, dict[str, Any]]:
    return load("spam_keywords.json").get("categories", {})


def generic_anchors() -> list[str]:
    return load("spam_keywords.json").get("generic_anchors", [])


def scoring() -> dict[str, Any]:
    return load("scoring.json")


def reset_cache() -> None:
    """Drop memoised config. Used by tests and after an admin edits a file."""
    with _lock:
        _cache.clear()

File: node-backend/domainMonitor/configLoader.js
"use strict";

// Faithful Node port of backend/domain_monitor/config_loader.py
const fs = require("fs");
const path = require("path");

const CONFIG_DIR =
  process.env.DOMAIN_MONITOR_CONFIG || path.join(__dirname, "config");

const _cache = {};

/**
 * Read a JSON config file once and memoise it. Config is data, not code.
 */
function load(name) {
  if (Object.prototype.hasOwnProperty.call(_cache, name)) return _cache[name];
  const p = path.join(CONFIG_DIR, name);
  let data;
  try {
    data = JSON.parse(fs.readFileSync(p, "utf-8"));
  } catch (exc) {
    console.error(`Could not read config ${p}: ${exc}`);
    data = {};
  }
  _cache[name] = data;
  return data;
}

function topics() {
  return load("seo_topics.json").topics || {};
}

function spamCategories() {
  return load("spam_keywords.json").categories || {};
}

function genericAnchors() {
  return load("spam_keywords.json").generic_anchors || [];
}

function scoring() {
  return load("scoring.json");
}

function resetCache() {
  for (const k of Object.keys(_cache)) delete _cache[k];
}

module.exports = {
  CONFIG_DIR,
  load,
  topics,
  spamCategories,
  genericAnchors,
  scoring,
  resetCache,
};
